### PR TITLE
add support for "systemctl --user --machine=foobar@.host" for connecting to user bus of user "foobar"

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/basic/hostname-util.h [hunk 1]
```diff
@@ -14,10 +14,14 @@ char* gethostname_short_malloc(void);
 int gethostname_strict(char **ret);
 
 bool valid_ldh_char(char c) _const_;
-bool hostname_is_valid(const char *s, bool allow_trailing_dot) _pure_;
-char* hostname_cleanup(char *s);
 
-#define machine_name_is_valid(s) hostname_is_valid(s, false)
+typedef enum ValidHostnameFlags {
+        VALID_HOSTNAME_TRAILING_DOT = 1 << 0,   /* Accept trailing dot on multi-label names */
+        VALID_HOSTNAME_DOT_HOST     = 1 << 1,   /* Accept ".host" as valid hostname */
```
Comment (dnicolodi) [inline src/basic/hostname-util.h:20]: Not that it matters much, but `VALID_HOSTNAME_TRAILING_DOT` and `VALID_HOSTNAME_LEADING_DOT` would be a  more consistent way of naming these flags.

Comment (poettering) [reply src/basic/hostname-util.h:20]: no. it's not the trailing dot we accept. it's literally ".host" we accept as reference for the local system. i.e. the 5 characters: dot, "h", "o", "s", "t".

Comment (dnicolodi) [reply src/basic/hostname-util.h:20]: I see. Sorry for the noise.

File: src/basic/hostname-util.c [hunk 2]
```diff
@@ -98,28 +98,24 @@ bool valid_ldh_char(char c) {
...
-bool hostname_is_valid(const char *s, bool allow_trailing_dot) {
+bool hostname_is_valid(const char *s, ValidHostnameFlags flags) {
         unsigned n_dots = 0;
         const char *p;
         bool dot, hyphen;
 
+        /* Check if s looks like a valid hostname or FQDN. This does not do full DNS validation, but only
+         * checks if the name is composed of allowed characters and the length is not above the maximum
+         * allowed by Linux (c.f. dns_name_is_valid()). Trailing dot is allowed if
+         * VALID_HOSTNAME_TRAILING_DOT flag is set an at least two components are present in the name. Note
```
Comment (bluca) [inline src/basic/hostname-util.c]: typo: an -> and

File: src/libsystemd/sd-bus/sd-bus.c [hunk 3]
```diff
@@ -1514,44 +1515,227 @@ _public_ int sd_bus_open_system_remote(sd_bus **ret, const char *host) {
...
+        /* Checks if a container specification in the form "user@container" or just "container" is valid.
+         *
+         * If the "@" syntax is used we'll allow either the "user" or the "container" part to be omitted, but
+         * not both. */
+
+        h = strchr(user_and_machine, '@');
+        if (!h)
+                h = user_and_machine;
+        else {
+                _cleanup_free_ char *b = NULL;
```
Comment (bluca) [inline src/libsystemd/sd-bus/sd-bus.c]: nit: s/b/user/ for readability

File: src/libsystemd/sd-bus/sd-bus.c [hunk 4]
```diff
@@ -1514,44 +1515,227 @@ _public_ int sd_bus_open_system_remote(sd_bus **ret, const char *host) {
...
+                            "argv4=-pUser%3d", eu ?: "root", ",",
+                            "argv5=-pPAMName%3dlogin,"
+                            "argv6=systemd-stdio-bridge");
+                if (!a)
+                        return -ENOMEM;
+
+                if (user) {
+                        char *k;
+
+                        /* Ideally we'd user the "--user" switch, but it's only available in more recent
```
Comment (bluca) [inline src/libsystemd/sd-bus/sd-bus.c]: typo: s/we'd user/we'd use/

**Review by bluca — APPROVED** [review]
Went through the diff and it looks good, just a couple of nits.

**Comment by bluca** [comment]
Do we have any test coverage of systemctl --machine?

File: src/libsystemd/sd-bus/bus-control.c [hunk 5]
```diff
@@ -713,7 +713,7 @@ _public_ int sd_bus_get_name_creds(
                 }
 
                 r = bus_creds_add_more(c, mask, pid, 0);
-                if (r < 0)
+                if (r < 0 && r != -ESRCH) /* If the process is gone by now, then don't augment, return what we got */
```
Comment (keszybz) [inline src/libsystemd/sd-bus/bus-control.c]: I'd prefer to have the comment reversed: "Return the error, but ignore ESRCH which just means the process is already gone" ...

File: src/libsystemd/sd-bus/sd-bus.c [hunk 6]
```diff
@@ -1514,44 +1515,227 @@ _public_ int sd_bus_open_system_remote(sd_bus **ret, const char *host) {
...
-        e = bus_address_escape(machine);
-        if (!e)
-                return -ENOMEM;
+        rhs = strchr(machine, '@');
+        if (rhs || user) {
+                _cleanup_free_ char *u = NULL, *eu = NULL, *erhs = NULL;
+
+                /* If there's an "@" in the container specification, we'll connect as a user specified at its
+                 * left hand side, which is useful in combination with user=true. This isn't as trivial as it
+                 * might sounds: it's not sufficient to enter the container and connect to some socket there,
```
Comment (keszybz) [inline src/libsystemd/sd-bus/sd-bus.c]: "might sounds" → "sounds" or "might sound"

File: src/libsystemd/sd-bus/sd-bus.c [hunk 7]
```diff
@@ -1514,44 +1515,227 @@ _public_ int sd_bus_open_system_remote(sd_bus **ret, const char *host) {
...
+                            "argv6=systemd-stdio-bridge");
+                if (!a)
+                        return -ENOMEM;
+
+                if (user) {
+                        char *k;
+
+                        /* Ideally we'd user the "--user" switch, but it's only available in more recent
+                         * systemd-stdio-bridge versions. Use the "-p" switch works too and is compatible
+                         * with older versions too, hence let's use that. */
```
Comment (keszybz) [inline src/libsystemd/sd-bus/sd-bus.c]: Strange sentence here.

**Review by keszybz — COMMENTED** [review]
Looks nice.

**Comment by poettering** [comment]
Force  pushed new version. Since changes where only renaming vars, and fixing up comments taking liberty to upgrade green label.

**Comment by poettering** [comment]
> Do we have any test coverage of systemctl --machine?

We do not. :-(

**Pull request merged**